### ISSUE TITLE
Execute Javascript is throwing an error

### ISSUE THREAD

**Issue by WCByrne** [skpm/sketch-module-web-view#111]
After updating from `3.0.1` -> `3.0.7` I'm getting an error when calling executeJavascript. Looks like the issue disappears if I roll back to `3.0.2`. `3.0.3+` doesn't work.

Here is how I'm calling it, pretty basic:
```js
window.webContents.executeJavaScript(script).catch(console.error);
```

Without changing anything except a different version `3.0.3+` of the web view, calling ☝️   now reaches an error on [this line](https://github.com/skpm/sketch-module-web-view/blob/ea2effe1c94a3377ef55dfed4669d3c9ed5be872/lib/execute-javascript.js#L12). Something with the state check:
```js
    // if the webview is not ready yet, defer the execution until it is
    if (webview.navigationDelegate().state.wasReady == 0) {
      return new Promise(function(resolve, reject) {
```

**Comment by mathieudutour** [comment]
What's the error?

**Comment by WCByrne** [comment]
That would be helpful wouldn't it 🤦‍♂. I won't be able to double check this until next week but IIRC it was an `undefined is not a function` type error. Something about the delegate or `.state` getter.

**Comment by WCByrne** [comment]
Here are the errors I'm seeing with the various versions.

## 3.0.2
No error

## 3.0.3:
```
TypeError: null is not an object (evaluating 'webview.navigationDelegate().valueForKey('state').wasReady')
executeJavaScript execute-javascript.js:12:0
postMessage webview.js:18:2
index.js:56:8
index.js:72:0
timeout.js:18:0
```

## 3.1.0:
```
MOJavaScriptException: TypeError: undefined is not an object (evaluating 'webview.navigationDelegate().state.wasReady')
executeJavaScript execute-javascript.js:12:0
postMessage webview.js:18:2
notifyDocumentSaved webview.js:104:14
onDocumentSave index.js:20:2
__skpm_run index.js:10870:17
__skpm_run [native code]
```

**Comment by mathieudutour** [comment]
for 3.0.3 it's "expected", and it was fixed right after: https://github.com/skpm/sketch-module-web-view/commit/a00a678cbf9e3f27a38919f9c68cdf930d1f60f5

Could you add a `console.log(webview.navigationDelegate())`? It seems that somehow the navigation delegate of the webview isn't the one we created 🤔

**Comment by WCByrne** [comment]
Yup, it's different 😕 

### 3.1.0
```
MochaJSDelegate_DynamicClass_ED03200A
```
### 3.0.2
```
NavigationDelegateClass2C051E66
```

**Comment by mathieudutour** [comment]
Well it's expected that it's a MochaJSDelegate class now. But I'm not sure why there isn't `state` available on it 🤔 

Could you log `webview.navigationDelegate().class().mocha().properties()` and `webview.navigationDelegate().class().mocha().instanceMethods ()`?

**Comment by WCByrne** [comment]
Very strange, it will run fine for a few seconds (5-10) after launching. delegate.state is an object. Then for some reason state becomes null. It is still in the properties but is no longer an object.

My only other idea so far is that I'm loading the window using `BrowserWindow.fromId`

And the only customizations I have made to the window when creating it are:
```
window._panel.setHidesOnDeactivate(false);
  window.setVisibleOnAllWorkspaces(true);
  window._panel.setFrameAutosaveName("com.lingoapp.sketch.windowFrame");
  window._panel.setFrameUsingName_force(
    window._panel.frameAutosaveName(),
    false
  );
```

**Comment by WCByrne** [comment]
Could this by chance be a 10.15 issue? I haven't seen it for a while on my 10.14 machine but am getting it again with 10.15.

**Comment by WCByrne** [comment]
Another detail I just noticed. I post messages to the webview in 2 ways:
1. In response to sketch actions like selectionChanged
2. To "callback" in response to messages the webview posted to the plugin. Something like insert get symbols, the plugin code gathers symbol data then posts a message back with that data.

It seems like the second type work fine while those in response to actions do not. Do the actions call code in a way that could cause an issue in the web view?

**Comment by mathieudutour** [comment]
That's 2 different code paths so it's possible that one is broken: specifically 1. is using the `remote` thing. It might be related to https://github.com/skpm/sketch-module-web-view/issues/123

**Comment by WCByrne** [comment]
Sounds very similar. I have a `webview.js` file which does just about the same thing mentioned in that issue. This is essentially a wrapper around the webview:

```js
const postMessage = (type, payload) => {
  const window = getWindow();
  if (!window) {
    console.error(
      "Attempt to post message but the Lingo window doesn't exist yet"
    );
    return;
  }
  const data = JSON.stringify({ type, payload });
  const script = `sketchBridge(${data})`;
  window.webContents.executeJavaScript(script).catch(console.error);
};

const createWindow = () => {
  const window = BrowserWindow({
    identifier: windowIdentifier,
    show: false,
    minWidth: 540,
    minHeight: 300
  });

  window._panel.setHidesOnDeactivate(false);
  window.setVisibleOnAllWorkspaces(true);
  window._panel.setFrameAutosaveName("com.lingoapp.sketch.windowFrame");
  window._panel.setFrameUsingName_force(
    window._panel.frameAutosaveName(),
    false
  );

  window.loadURL(require("../resources/webview.html")); // eslint-disable-line
  _injectSettings(window);
  return window;
};
```

Since my plugin only ever has one window `createWindow` is called once when the plugin loads then `postMessage` is used to communicate to that window.

**Comment by mathieudutour** [comment]
if you only need to send a message to the window, you should probably use `require('sketch-module-web-view/remote').sendToWebview(identifier, script)`, it will be much faster. (but we should still fix this)

**Comment by mathieudutour** [comment]
also only requiring `sketch-module-web-view/remote` when you can will make your commands a lot smaller

**Comment by mathieudutour** [comment]
`window._panel.setHidesOnDeactivate(false);` that's good! I'm stealing it 😄

**Comment by mathieudutour** [comment]
```js
window._panel.setFrameAutosaveName("com.lingoapp.sketch.windowFrame");
  window._panel.setFrameUsingName_force(
    window._panel.frameAutosaveName(),
    false
  );
```
Do you think that should be the default (using the given identifier)?

**Comment by WCByrne** [comment]
I'll give the `/remote` functions a try. Didn't know that existed.

Would be nice if both of those were options on BrowserWindow. Maybe even save the frame by default.
```
const window = BrowserWindow({
    identifier: "unique.id",
    hideOnDeactivate: false, // default to true for existing behavior?
    saveFrame: true
  });
```

**Comment by WCByrne** [comment]
sooo.... replaced `getWindow().webContents.executeJavaScript(script)` with `sendToWebview` and I'm not seeing the navigationDelegate errors anymore. 🎉

Not sure if there is still some underlying issue that should be addressed but that works for me!

**Comment by mathieudutour** [comment]
fixed in 3.2.0
